controller/contour/rbac.go

```diff
@@ -0,0 +1,263 @@
+package contour
+
+import (
+	"context"
+	"fmt"
+	operatorv1alpha1 "github.com/projectcontour/contour-operator/api/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	rbacv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+const (
+	defaultContourServiceAccount = "contour"
+	defaultEnvoyServiceAccount   = "envoy"
+)
+
+// ensureRBAC ensures all the necessary RBAC resources exist for the
+// provided contour.
+func (r *Reconciler) ensureRBAC(ctx context.Context, contour *operatorv1alpha1.Contour) error {
+	ns := contour.Spec.Namespace.Name
+	ctr := types.NamespacedName{
+		Namespace: ns,
+		Name:      defaultContourServiceAccount,
+	}
+	envoy := types.NamespacedName{
+		Namespace: ns,
+		Name:      defaultEnvoyServiceAccount,
+	}
+	names := []types.NamespacedName{ctr, envoy}
+	for _, name := range names {
+		if err := r.ensureServiceAccount(ctx, name.Namespace, name.Name); err != nil {
+			return fmt.Errorf("failed to create service account %s/%s: %w", name.Namespace, name.Name, err)
```

Comment:
@Miciah I'm getting ready to push an update that looks like:
```
	for _, name := range names {
		if err := r.ensureServiceAccount(ctx, name.Namespace, name.Name); err != nil {
			return fmt.Errorf("failed to ensure service account for contour %s/%s: %w", contour.Namespace,
				contour.Name, err)
		}
	}
```
This particular error now references the Contour object instead of the ServiceAccount namespace/name. `err` will reference the details of the `ensureServiceAccount()` error. I do like using ns/name since `Contour` and `ServiceAccount` are namespace-scoped.

Reply:
Printing the stringified `types.NamespacedName` value is simpler if you have a `types.NamespacedName` value, but yeah, this doesn't apply with a `*operatorv1alpha1.Contour` value.